docs/features/incremental-generators.md

```diff
@@ -0,0 +1,485 @@
+    public static IncrementalValueSource<(T, IEnumerable<U>)> Combine<T, U>(this IncrementalValueSource<T> source1, IncrementalValueSource<U> source2) => ...
+}
+```
+
+The second data source is batched, and the resulting step has a value type of
+`(T, IEnumerable<U>)`. That is, a tuple where each element consists of a single
+item of data from `source1` combined with the entire batch of data from
+`source2`. While this is somewhat low-level, when combined with subsequent
+transforms, it gives the author the ability to combine an arbitrary number of
+data sources into any shape they require.
+
+In the following example the author combines the additional text source with the
+compilation:
+
+```csharp
+var combined = context.Sources.AdditionalTexts.Combine(context.Sources.Compilation);
+```
+
+The type of combined is an
+`IncrementalValueSource<(AdditionalText, IEnumerable<Compilation>)>`. For each
+additional text, there is a tuple with the text, and the batched data of the
+compilation source. As the author knows that there is only ever a single
+compilation, they are free to transform the data to select the single
+compilation object:
+
+```csharp
+IncrementalValueSource<(AdditionalText, Compilation)> transformed = combined.Transform(pair => (pair.Item1, pair.Item2.Single()));
+```
+
+Similarly a cross join can be achieved by first combining two value sources,
+then batch transforming the resulting value source.
+
+```csharp
+IncrementalValueSource<(AdditionalText, MetadataReference)> combined = context.Sources.AdditionalTexts
+                                                                                      .Combine(context.Sources.MetadataReference)
+                                                                                      .TransformMany(pair => pair.Item2.Select(item2 => (pair.Item1, item2));
+```
+
+**OPEN QUESTION**: Combine is pretty low level, but does allow you to do
+everything needed when used in conjunction with transform. Should we provide
+some higher level extension methods that chain together combine and transform
+out of the box for common operations?
+
+While filtering can be easily enough implement by the user as a transform step,
+it seems common and useful enough that we provide an implementation directly for
+the user to consume
+
+```csharp
+static partial class IncrementalValueSourceExtensions
+{
+    // helper for filtering values
+    public static IncrementalValueSource<T> Filter<T>(this IncrementalValueSource<T> source, Func<T, bool> filter) => ...
+}
+```
+
+### Caching
+
+While the finer grained steps allow for some coarse control of output types via
+the generator host, the performance benefits are only really seen when the
+driver can cache the outputs from one pipeline step to the next. While we have
+generally said that the execute method in an `ISourceGenerator` should be
+deterministic, incremental generators actively _require_ this property to be
```

Comment:
What failure modes do you imagine will occur when developers get this wrong? Should there be a "debug/checked" mode of some kind that incremental generators can run in so that issues around nondeterminism in user code can be caught more easily?